internals/controllers/updatequota/updatequota_controller.go

```diff
@@ -0,0 +1,349 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// UpdateQuotaReconciler reconciles a UpdateQuota object
+type UpdateQuotaReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+const maxRetries = 600
+
+// +kubebuilder:rbac:groups=dana.hns.io,resources=updatequota,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=dana.hns.io,resources=updatequota/status,verbs=get;update;patch
+
+func (r *UpdateQuotaReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&danav1.Updatequota{}).
+		Complete(r)
+}
+
+func (r *UpdateQuotaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName("controllers").WithName("UpdateQuota").WithValues("upq", req.NamespacedName)
+	logger.Info("starting to reconcile")
+
+	upqObject, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, &danav1.Updatequota{})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get object '%s': "+err.Error(), upqObject.Object.GetName())
+	}
+
+	if !upqObject.IsPresent() {
+		logger.Info("resource not found. Ignoring since object must be deleted")
+		return ctrl.Result{}, nil
+	}
+
+	phase := upqObject.Object.(*danav1.Updatequota).Status.Phase
+	if utils.ShouldReconcile(phase) {
+		return ctrl.Result{}, r.reconcile(upqObject)
+	} else {
+		logger.Info("no need to reconcile, object phase is: ", "phase", phase)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *UpdateQuotaReconciler) reconcile(upqObject *utils.ObjectContext) error {
+	ctx := upqObject.Ctx
+	logger := log.FromContext(ctx)
+
+	sourceNSName := upqObject.Object.(*danav1.Updatequota).Spec.SourceNamespace
+	sourceNS, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Name: sourceNSName}, &corev1.Namespace{})
+	if err != nil {
+		return fmt.Errorf("failed to get object '%s': "+err.Error(), sourceNSName)
+	}
+
+	destNSName := upqObject.Object.(*danav1.Updatequota).Spec.DestNamespace
+	destNS, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Name: destNSName}, &corev1.Namespace{})
+	if err != nil {
+		return fmt.Errorf("failed to get object '%s': "+err.Error(), destNSName)
+	}
+
+	// get the Ancestor namespace of the source and destination namespaces. The Ancestor namespace is the
+	// first namespace the two namespaces have in common in their hierarchy. There are several cases and
+	// each case is treated differently based on the Ancestor namespace
+	sourceNSArray := utils.GetNSDisplayNameArray(sourceNS)
+	destNSArray := utils.GetNSDisplayNameArray(destNS)
+
+	ancestorNSName, _, err := utils.GetAncestor(sourceNSArray, destNSArray)
+	if err != nil {
+		return fmt.Errorf("failed to find ancestor namespace of '%s' and '%s': "+err.Error(), sourceNSArray, destNSArray)
+	}
+
+	if sourceNS.Object.GetName() == ancestorNSName {
```

Comment:
ill divide the functions of update quota one until the else and explain what the function does

Reply:
divided